### ISSUE TITLE
The result is not deterministic when infer same input.

### ISSUE THREAD

**Issue by xinzheshen** [fatchord/WaveRNN#165]
Hello, when I infer the same input in my modified version [code](https://github.com/xinzheshen/tacotron2/blob/1c8970a85d289288de6ec1030e55dc65981e623f/wavernn/fatchord_version.py#L221), the result is not deterministic sometimes.
And I have set the random seed at the beginning as follow.

`def setup_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.deterministic = True`

Do you have any suggestions? Thank you.

**Comment by mindmapper15** [comment]
https://github.com/Rayhane-mamah/Tacotron-2/issues/155#issuecomment-413364857

https://github.com/ibab/tensorflow-wavenet/issues/347

Yeah, That's totally normal.
Even if you use softmax rather than MoL, random sampling from softmax distribution has better results than choosing argmax.

**Comment by xinzheshen** [comment]
Thank you @mindmapper15 . I got it a little. and still confused.
For example, when I execute code below everytime, it has same result. So the distribution and random seed are fixed, the result should be same, isn't it?
`  seed = 1234
   torch.manual_seed(seed)
   torch.cuda.manual_seed(seed)

   m = torch.distributions.Categorical(torch.tensor([0.25, 0.25, 0.4, 0.1]).cuda())
   print(m.sample())
   for i in range(10):
           print(m.sample())  
`
`output:
tensor(0, device='cuda:0')
tensor(2, device='cuda:0')
tensor(0, device='cuda:0')
tensor(2, device='cuda:0')
tensor(0, device='cuda:0')
tensor(2, device='cuda:0')
tensor(3, device='cuda:0')
tensor(2, device='cuda:0')
tensor(2, device='cuda:0')
tensor(0, device='cuda:0')
`

**Comment by mindmapper15** [comment]
I'm not sure what the cause is...
But I found something interesting comment on the other repository.

https://github.com/pytorch/pytorch/issues/7068#issuecomment-484918113

Maybe you should checkout!
I hope you solve the problem :)

**Comment by xinzheshen** [comment]
@mindmapper15 Thank you :)

**Comment by 742617000027** [comment]
Try setting
```
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```
as suggested by this article on [Reproducibility in PyTorch](https://pytorch.org/docs/stable/notes/randomness.html).
